### PR TITLE
Make sure lastAppProps always have some value.

### LINKED ISSUE #827: Server-side error in with-loading example?
The [with-loading](https://github.com/zeit/next.js/tree/master/examples/with-loading) example works well client-side, but on hard refreshing the `/forever` or `/about` routes, it throws `Cannot read property 'Component' of undefined`. `Component` is being imported from `React` in those two pages. Full trace:

```
TypeError: Cannot read property 'Component' of undefined
    at _callee3$ (http://localhost:3000/_next/-/main.js:11570:50)
    at tryCatch (http://localhost:3000/_next/-/commons.js:12871:40)
    at Generator.invoke [as _invoke] (http://localhost:3000/_next/-/commons.js:13162:22)
    at Generator.prototype.(anonymous function) [as next] (http://localhost:3000/_next/-/commons.js:12923:21)
    at step (http://localhost:3000/_next/-/commons.js:3111:30)
    at http://localhost:3000/_next/-/commons.js:3129:14
    at F (http://localhost:3000/_next/-/commons.js:1634:28)
    at http://localhost:3000/_next/-/commons.js:3108:12
    at doRender (http://localhost:3000/_next/-/main.js:11587:18)
    at Object._callee$ (http://localhost:3000/_next/-/main.js:11489:20)
```

### PR DESCRIPTION
**Pull request by arunoda**
Fixes #827

### CODE REVIEW

File: client/index.js [hunk 1]
```diff
@@ -20,7 +20,7 @@ const {
 
 const Component = evalScript(component).default
 const ErrorComponent = evalScript(errorComponent).default
-let lastAppProps
```
Comment (arunoda) [inline client/index.js]: When the props became `null` our code tries to read the props from `lastAppProps`.

Then, For the very first doRender call, it tries to check few items from this `lastAppProps`. Unfortunately, it's a null and we'll get the mentioned problem.

**Comment by nkzawa** [comment]
This wouldn't fix #827 completely, since `getInitialProps` have to return an object. We should set an empty object as default value if `getInitialProps` didn't return anything, or display a warning.

**Comment by timneutkens** [comment]
I can't imagine a case Where you'd want to make getInitialProps a void. So it might be best to show an error when the promise Doesn't resolve to an object. 🤔

**Comment by arunoda** [comment]
@nkzawa see this line https://github.com/arunoda/next.js/blob/b4ae722d9c1a4460e17dbdc041b111cbd492b2aa/client/index.js#L70

**Comment by nkzawa** [comment]
@arunoda I meant  `props = props || lastAppProps.props` (https://github.com/arunoda/next.js/blob/b4ae722d9c1a4460e17dbdc041b111cbd492b2aa/client/index.js#L68).

We don't want to allow empty `props` here since reusing `lastAppProps` is for HMR.

**Comment by rauchg** [comment]
I also think we should display a warning upon dev, and default it. We should enforce strict types whenever possible.

**Comment by arunoda** [comment]
@rauchg yes. Now we throw an error like this:
/cc @nkzawa 

screen shot 2017-01-21 at 1 01 26 am

**Pull request merged**